### ISSUE TITLE
Question: Disposing on Different Thread

### ISSUE THREAD

**Issue by TheXenocide** [StephenCleary/AsyncDiagnostics#4]
Hello,

Background: I have a very similar (if slightly more complex) implementation of contextual diagnostic information that is retained across the execution context, though mine uses CallContext instead of AsyncLocal (since AsyncLocal didn't exist yet at the time it was implemented). One of the issues that can lead to a memory leak is if the contextual object is disposed on a different thread from the one it was created on. This will pop it from the stack of the thread it's on, but the stack remains unchanged on the original thread's CallContext. If reusing this code repetitively the stack simply continues to grow. This has become more challenging to manage in the world of async/await since code much deeper down the stack than the place where the implicit context is created can cause thread context to switch, so the location in which the using statement exists has no control. We throw an exception if disposing the object on a different thread so we can detect it, but it's still a pretty severe limitation. We've recently started experiencing the safeguard exception we placed more frequently, partly after an update to .NET 4.7.2 which changed the behavior of some awaited activities in our REST API stack, as well as in a WCF operation invoker we use which hasn't typically shown any thread switching behavior (this could be related to our own code further down the stack now using await when it didn't previously; it's a large codebase so I haven't had a chance to review all changes that might possibly have impacted this particular service).

As for the question: I've been trying to understand the subtle differences between CallContext and AsyncLocal, both of which are carried forth by the ExecutionContext but appear to have slightly different wireup mechanics when migrating to a new thread. Do you know for sure that AsyncLocal is removed from the context of a thread that has been reclaimed if asynchronous execution further down the call stack causes the operation to be migrated to a different thread? Would switching our implementation to use AsyncLocal give us the same disposable pattern for establishing context around a logical block of code without having to worry about which physical threads they are created/disposed on, as long as it is disposed in a subsequent stage of the same logical flow of code?

I appreciated your article on the subject and thought you might know more about the subtle implementation detail differences between AsyncLocal and CallContext which are not particularly deeply documented out in the wild. Any insight you might be able to provide would be greatly appreciated.

Thanks.

**Comment by StephenCleary** [comment]
Yes, I believe `AsyncLocal` will have the semantics you need. The context does not stay stuck on threads.

There's a couple of different call contexts; there's the normal call context and the *logical* call context. These are really quite different. `AsyncLocal` is essentially a new and simplified API for the *logical* call context. The normal call context (humorously referred to as the "illogical call context" in the source) is tied to a thread; the logical call context is not.

There is an additional consideration when using `AsyncLocal<T>` / logical call context: it is now possible for the logical stack to grow much larger. Asynchronous recursive code doesn't grow on the stack; it grows on the heap. So it's possible to have very deeply nested asynchronous recursive code that could cause your diagnostic stack to grow larger than it could when bounded by a single thread's stack. This is not a concern for most code, but something worth noting.

Another consideration is one of performance: logical contexts including `AsyncLocal<T>` must propagate to other threads, and .NET is highly optimized for the "no logical context" scenario. Again, probably not a concern for most code (asynchronous code performance is generally driven by I/O, and adding a bit of CPU work doesn't make a measurable difference), but worth noting.

Finally, there's the consideration of scope. `async` methods establish a copy-on-write scope for the logical call context, which can cause some unexpected "sharing" of values. The scope will be shared for all methods that are in the same synchronous block. A few examples:

- `A() => B() => C()` (all synchronous) - all share the same scope. If `C` sets an `AsyncLocal<T>` value, it will be seen by `A`.
- `A() => async B() => C()` - `B` and `C` share the same scope. `A` is in a different scope and will not see `AsyncLocal<T>` writes from `B` or `C`.
- `async A() { B(); C(); }` - `A` establishes its own scope, which is shared by `B` and `C`. `C` will see `AsyncLocal<T>` writes from `B` even though it's not called by `B`.
- Context is cleared when threads return to the thread pool. So, even if a synchronous method `B` sets an `AsyncLocal<T>` value, and that same thread is later used to run a synchronous method `C`, `C` will still not see the write from `B` because that thread's context was cleared.

The scoping can get a bit confusing, but the problems are easily avoided by using the `Dispose` pattern, which you're already doing. When used in an `async` method, the `Dispose` will be doing unnecessary work (since the caller cannot see the updated value anyway), but it is necessary for reasonable scoping in synchronous methods.

TL;DR: Yes, `AsyncLocal<T>` when used with the `Dispose` pattern will produce the semantics you need.

**Comment by TheXenocide** [comment]
Thanks very kindly for the thorough reply. I'm personally already familiar with the logical and "illogical" contexts (our dated `CallContext` implementation has immutable context objects which implement `ILogicalThreadAffinative` and utilize `LogicalSetData`) but even that portion would hopefully be helpful to anybody who stumbles on this going forward. 

My concerns originally stemmed from the notion that we observe memory leaks if we store a stack on the logical call context unless we remove it from the very same thread. Your articles and this answer helped weigh my time and I have since done a small POC that verifies your findings with `AsyncLocal<T>` (I even manually migrated a Task from being created in one Task and awaited in another, which yielded pleasing results) so I'm happy to be able to propose a significant improvement to an area in our solution which has been troubled as `async`/`await`/`Task` see increased esoteric uses throughout our code. In case anybody else coming along here finds this disucssion interesting, I'll note that I have since stumbled on [this forum post](https://social.msdn.microsoft.com/Forums/en-US/ea21ca57-5340-439c-8ee9-f0185b5787a1/callcontext-what-are-the-recommendations-for-using-this-going-forwards?forum=async#a6d5ce0e-53bb-47e3-a3ee-d84d6f062dcb) (yet another of your contributions to the academics of the subject, apparently prior to the advent of `AsyncLocal<T>`) which describes a very similar type of situation to what we've been encountering. 

One last observation I'll share for others' consideration if they should find this discussion helpful is that, unlike `CallContext` with `ILogicalThreadAffinative`, `AsyncLocal<T>` does *not* participate in cross-`AppDomain` execution context exchange. This behavior simplifies implementation in some ways, since objects no longer need to supply the `[Serializable]` attribute or inherit from `MarhsalByRefObject` (which has the added complexity of lifetime management which has lead to memory leaks and other exceptions in some utilizations of this I've seen), but it also means that any implicit context which existing implementations might depend on traveling across `AppDomain` boundaries will need to consider that solution separately from the `async`/`await` solution this provides. 

It seems you and I both have had a long history with implicit context. It has been a pleasure to have shared this exchange and thanks for your contributions.

**Comment by StephenCleary** [comment]
Thanks for the follow-up! I wasn't aware of the semantics of `AsyncLocal<T>` with app domains.
